src/julia_gc.cc

```diff
@@ -210,204 +126,471 @@ static inline void * align_ptr(void * p)
     return (void *)u;
 }
 
-#if !defined(SCAN_STACK_FOR_MPTRS_ONLY)
 
-typedef struct treap_t {
-    struct treap_t *left, *right;
-    size_t          prio;
-    void *          addr;
-    size_t          size;
-} treap_t;
+// So that we don't have to link against the standard C++ library, we do
+// manual allocation via alloc() and dealloc() routines that have their own
+// error handling. Allocate instances of ArrayList and BalancedTree via
+// ArrayList<T>::make(...) and BalancedTree<T, f>::make(...).
+//
+// Likewise, ...::destroy(...) can be used to deallocate the data structures
+// and their contents. If placed in manually allocated memory, call init()
+// and deinit() to initialize and deinitialize them.
+//
 
-static treap_t * treap_free_list;
 
-treap_t * alloc_treap(void)
-{
-    treap_t * result;
-    if (treap_free_list) {
-        result = treap_free_list;
-        treap_free_list = treap_free_list->right;
+class StdAlloc {
```

Comment:
That said, this class `StdAlloc` feels like over-engineering to me. Can't we just have the global `alloc` template function call `malloc`? Or even simpler, just overload operator `new` and `delete`?
```C++
void* operator new(std::size_t sz)
{
    void *result = std::malloc(sz);
    if (!result)
        abort();
}
void operator delete(void* ptr) noexcept
{
    std::free(ptr);
}
```

Reply:
I tried custom `operator new` and `operator delete`, but that requires that the C++ runtime is present. This is currently not the case, because the code is linked with gcc instead of g++. If I could rely on the runtime, that would simplify things enormously, as I also wouldn't have to hack around missing constructor/destructor calls.

The existence of `StdAlloc` is in part a leftover from trying to have those operators, that I then reengineered to just encapsulate the allocation mechanism.

Reply:
As for C vs. C++, it's perfectly possible to redo this in C. The downside is that generics generally do not result in pretty or maintainable C code. You can either use macros to emulate generics or you can instantiate the generics from a template, e.g. as in `src/sortbase.h`. It's no big deal to rewrite it this way (probably something like a day of coding/testing), though, it's just a concern about code maintenance.

Reply:
I wasn't aware that operator new and delete need the C++ runtime, though I guess it makes some sense. Anyway, it is absolutely fine by me to not use that, I was just wondering... Thanks for the explanation.

As to redoing this in C: Actually, personally I am with keeping this in C++. Esp. if the code already existed before and is somewhat tested; a rewrite is just a chance to introduce new regressions :-).

Reply:
That said, I'd still recommend to get rid of `StdAlloc` as a base class, i.e., no inheritance; just provide global `alloc` / `dealloc` template functions which always call `malloc` / `free`

Reply:
Heh. I just went and redid it in C. As the rewrite was mostly just mechanical, I'm not too worried about regressions. That said, I'm personally fine either way.